nexus/src/app/background/inventory_collection.rs

```diff
@@ -224,18 +224,28 @@ mod test {
         );
 
         // Nexus starts the very background task that we're also testing
-        // manually here.  As a result, we should find a collection in the
-        // database before too long.  Wait for it so that after it appears, we
+        // manually here. As a result, we should find a collection in the
+        // database before too long. Wait for it so that after it appears, we
         // can assume the rest of the collections came from the instance that
         // we're testing.
+        //
+        // This test setup also contains 2 sled-agents. Each of those
+        // sled-agents will upsert themselves via nexus, which will trigger
+        // another collection.
+        //
+        // Therefore we would expect to have to wait for a total of 3
+        // collections to appear before we run manual activations.
+        // TODO: Why then are we only seeing two collections?
```

Comment:
I am stumped here. I saw that both upserts were triggering collections. But I can also see that when nexus gets a handoff it activates the collector again. That should be 3 collections and not 2.

Reply:
I think I understand this. On this branch we see 5 activations of the inventory collection bg task.

The first activation is when the task is created; tokio timeouts always fire immediately. This always fails pre-rack-setup with `Forbidden` (I think because the database hasn't been populated with roles, etc., needed for queries to work?), but I think trying to activate immediately is still correct because on a Nexus start with a populated DB this ought to work.

The second and third activations are racing each other. One is the activation Nexus fires [after it populates the built-in data](https://github.com/oxidecomputer/omicron/blob/6ed661734385d90f388e66c51400165caadce4cb/nexus/src/app/mod.rs#L466-L494); this is `tokio::spawn`'d as a part of [starting the internal Nexus server](https://github.com/oxidecomputer/omicron/blob/6ed661734385d90f388e66c51400165caadce4cb/nexus/test-utils/src/lib.rs#L1338-L1341); I strongly suspect this one basically always wins the race, but since it's `spawn`'d it technically is racing. The other is from the first `sled_upsert` added on this PR.

The fourth activation is the second `sled_upsert` added on this PR.

The fifth activation is the collection triggered after the RSS handoff.

When I run this branch, I see the first three collections fail: the first fails with `Forbidden` (as expected), and the second and third both fail with a DNS resolution error:

```
14:01:01.187Z WARN ca2462e3-3f7e-454f-9c45-2afce70a6863 (ServerContext): inventory collection failed
    background_task = inventory_collection
    error = failed to collect inventory: looking up MGS addresses: no record found for Query { name: Name("_mgs._tcp.control-plane.oxide.internal."), query_type: SRV, query_class: IN }
```

The fourth and fifth activations succeed, matching the 2 you're seeing here.

However, I'll claim that activations 2, 3, and 4 are all racing against internal DNS being populated. To test this, I added a 10 second sleep just before `nexus-test-utils` [populates internal DNS](https://github.com/oxidecomputer/omicron/blob/6ed661734385d90f388e66c51400165caadce4cb/nexus/test-utils/src/lib.rs#L773); with that sleep, activations 2, 3, and 4 _all_ fail with DNS resolution errors, and we only see 1 collection at this point instead of 2 or 3. The final collection triggered by RSS handoff always succeeds, because nexus-test-utils doesn't kick that off until after populating internal DNS (and other things), so even with my artificial sleep this one is fine.

We could make this deterministic by reordering the init steps in the test context; e.g., by moving [`populate_internal_dns`](https://github.com/oxidecomputer/omicron/blob/6ed661734385d90f388e66c51400165caadce4cb/nexus/test-utils/src/lib.rs#L1362-L1365) up before [`start_sled1`](https://github.com/oxidecomputer/omicron/blob/6ed661734385d90f388e66c51400165caadce4cb/nexus/test-utils/src/lib.rs#L1342-L1349). I'm not sure if that's reasonable or not. In real RSS, if a sled is able to _find_ Nexus to register itself, that means DNS is up and the triggered inventory collection should work. But other parts of sleds coming online have to work even before internal DNS is up, since part of "sleds coming online" is bringing up internal DNS.

I think I'd be inclined to actually leave this test the way it was, and maybe test that sled_upsert triggers collections separately (e.g., via the add sled flow?).

Reply:
John and I talked a bunch about this in chat and agree that there will pretty much always be a race. John implemented a patch that tolerates races in this test that I just pushed up.